src/Microsoft.AspNet.Mvc/MvcServiceCollectionExtensions.cs

```diff
@@ -34,10 +34,77 @@ public static IServiceCollection AddMvc([NotNull] this IServiceCollection servic
+        /// discovery.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="controllerTypes">A sequence of controller <see cref="Type"/>s to register in the
+        /// <paramref name="services"/> and used for controller discovery.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+           [NotNull] this IServiceCollection services,
+           [NotNull] IEnumerable<Type> controllerTypes)
+        {
+            var controllerTypeProvider = new FixedSetControllerTypeProvider();
+            foreach (var type in controllerTypes)
+            {
+                services.TryAdd(ServiceDescriptor.Transient(type, type));
+                controllerTypeProvider.ControllerTypes.Add(type.GetTypeInfo());
+            }
+
+            services.Replace(ServiceDescriptor.Transient<IControllerActivator, ServiceBasedControllerActivator>());
+            services.Replace(ServiceDescriptor.Instance<IControllerTypeProvider>(controllerTypeProvider));
+
+            return services;
+        }
+
+        /// <summary>
+        /// Registers controller types from the specified <paramref name="assemblies"/> as services and as a source
+        /// for controller discovery.
+        /// </summary>
+        /// <param name="services">The <see cref="IServiceCollection"/>.</param>
+        /// <param name="controllerAssemblies">Assemblies to scan.</param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        public static IServiceCollection WithControllersAsServices(
+            [NotNull] this IServiceCollection services,
+            [NotNull] IEnumerable<Assembly> controllerAssemblies)
+        {
+            var assemblyProvider = new FixedSetAssemblyProvider();
+            foreach (var assembly in controllerAssemblies)
+            {
+                assemblyProvider.CandidateAssemblies.Add(assembly);
+            }
+
+            var controllerTypeProvider = new DefaultControllerTypeProvider(assemblyProvider);
+            var controllerTypes = controllerTypeProvider.ControllerTypes;
+
+            return WithControllersAsServices(services, controllerTypes.Select(type => type.AsType()));
+        }
+
+        // To enable unit testing
+        internal static void AddMvcServices(IServiceCollection services)
+        {
             // Options and core services.
-            services.TryAdd(ServiceDescriptor.Transient<IConfigureOptions<MvcOptions>, MvcOptionsSetup>());
-            services.TryAdd(
-                ServiceDescriptor.Transient<IConfigureOptions<RazorViewEngineOptions>, RazorViewEngineOptionsSetup>());
+            services.AddTransient<IConfigureOptions<MvcOptions>, MvcOptionsSetup>();
+            services.AddTransient<IConfigureOptions<RazorViewEngineOptions>, RazorViewEngineOptionsSetup>();
```

Comment:
This is fixing the following issue. Sure, will add a comment.
https://github.com/aspnet/Mvc/issues/2648